### ISSUE TITLE
Question: Why is CFG not set during training but set during inference?

### ISSUE THREAD

**Issue by uk9921** [guoqincode/AnimateAnyone-unofficial#14]
Operations such as `cfg_random_null_text` or `cfg_random_null_ref` were not used during the training phase, but `guidance_scale: 7.5` and `do_classifier_free_guidance=True` was set during inference. Is this as expected?

**Comment by guoqincode** [comment]
feel free to modify the Dataset or train policy, the current version is only a basic implementation

**Comment by uk9921** [comment]
Thanks for the reply. Is there any progress in your experiment? I can only generate very scary images now.
![image](https://github.com/guoqincode/AnimateAnyone-unofficial/assets/20817396/7b5063ab-a6de-4c0f-8bf8-f344dd25fd47)

**Comment by guoqincode** [comment]
HMMM... Something seems wrong. Can you share some of your training information with me?

**Comment by uk9921** [comment]
I used the main code you gave me and ran it on 5k videos for over 50 epochs. I’m trying hard to figure out why it went wrong. I also want to check if your tests worked out well.

**Comment by guoqincode** [comment]
Did you use cfg_random_null_ref during training? 
I previously worked with xformers and mixed precision when training on RTX 3090/4090 Gpus and there was a problem with parameters not updating. If you are an RTX series GPU, I recommend turning off xformers and mixed precision training.

**Comment by uk9921** [comment]
No, actually, I have no idea how to enable cfg_random_null_ref for image embedding during training and I also disable it in the inference pipeline.
And I will try to disable the xformers and fp16, thanks for your advice！

mark the result :D
![image](https://github.com/guoqincode/AnimateAnyone-unofficial/assets/20817396/f03c1625-1d47-42bc-95c7-4a35d462f379)

**Comment by guoqincode** [comment]
Now that I can get the normal results of the first phase of training, I suggest that you can test your training on a video to see if it is normal. In the meantime, I recommend using float32 for both reasoning and training.

**Comment by guoqincode** [comment]
If you are Chinese, you can send me an email: guoqin@stu.pku.edu.cn, we can add wechat.

**Comment by uk9921** [comment]
Why not? Let's wechat :D

**Comment by zhenzhiwang** [comment]
@uk9921 Have you solved this problem? I also encountered this problem when I trying to train a model in UBC dataset in stage 1. Is the reason is xformer and mixed precision training?

**Comment by uk9921** [comment]
@zhenzhiwang Hi, My problem is not enough training. You can try to turn off xformer and mixed precision and simply train to verify the difference.

**Comment by zhenzhiwang** [comment]
> @zhenzhiwang Hi, My problem is not enough training. You can try to turn off xformer and mixed precision and simply train to verify the difference.

Hi @uk9921, thanks for your quick reply.  I tested the 20000 iteration ckpt in UBC dataset and get similar images with https://github.com/guoqincode/AnimateAnyone-unofficial/issues/14#issuecomment-1855521920. Is it normal to get such meaningless images and suddenly become better in 30000 iterations (since that animate anyone paper reports that they only train 30k iterations).

**Comment by uk9921** [comment]
Hi @zhenzhiwang,  honestly it's hard to say, so many setup details and I have little experience with them. Data set, lr, epochs, modules to be trained, and lr of each module will significantly affect the final result.

Besides training, you can also debug the inference pipeline and make sure the inputs are pre-processed as expected. For example, after resize&crop, are there some hard input case such as half-body reference and full-body poses?
